crates/pallet-domains/src/block_tree.rs

```diff
@@ -0,0 +1,504 @@
+                        DomainBlocks::<T>::remove(block);
+                    }
+                    // Remove the block's `ExecutionInbox` as the block is pruned and not need
+                    // to verify the its receipt's `extrinsics_root` anymore
+                    ExecutionInbox::<T>::remove(domain_id, to_prune);
+                }
+            }
+        }
+        ReceiptType::CurrentHead => {
+            // Add confirmation to the current head receipt
+            DomainBlocks::<T>::mutate(er_hash, |maybe_domain_block| {
+                let domain_block = maybe_domain_block.as_mut().expect(
+                    "The domain block of `CurrentHead` receipt is checked to be exist in `execution_receipt_type`; qed"
+                );
+                domain_block.operator_ids.push(submitter);
+            });
+        }
+        // Other types of receipt is unexpected for this function
+        _ => return Err(Error::UnexpectedReceiptType),
+    }
+    Ok(())
+}
+
+/// Import the genesis receipt to the block tree
+pub(crate) fn import_genesis_receipt<T: Config>(
+    domain_id: DomainId,
+    genesis_receipt: ExecutionReceiptOf<T>,
+) {
+    let er_hash = genesis_receipt.hash();
+    let domain_block_height = genesis_receipt.domain_block_height;
+    let domain_block = DomainBlock {
+        execution_receipt: genesis_receipt,
+        operator_ids: sp_std::vec![],
+    };
+    // NOTE: no need to upate the head receipt number as we are using `ValueQuery`
+    BlockTree::<T>::mutate(domain_id, domain_block_height, |er_hashes| {
+        er_hashes.try_insert(er_hash)
+            .expect(
+                "Must not exceed MaxBlockTreeFork as the genesis receipt is the first and only receipt at block #0; qed"
+            );
+    });
+    DomainBlocks::<T>::insert(er_hash, domain_block);
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::domain_registry::DomainConfig;
+    use crate::tests::{
+        create_dummy_bundle_with_receipts, create_dummy_receipt, new_test_ext,
+        GenesisStateRootGenerater, ReadRuntimeVersion, Test,
+    };
+    use crate::NextDomainId;
+    use frame_support::dispatch::RawOrigin;
+    use frame_support::traits::{Currency, Hooks};
+    use frame_support::weights::Weight;
+    use frame_support::{assert_err, assert_ok};
+    use frame_system::Pallet as System;
+    use sp_core::H256;
+    use sp_domains::{GenesisReceiptExtension, RuntimeType};
+    use sp_runtime::traits::BlockNumberProvider;
+    use sp_version::RuntimeVersion;
+    use std::sync::Arc;
+
+    fn run_to_block<T: Config>(block_number: T::BlockNumber, parent_hash: T::Hash) {
+        System::<T>::set_block_number(block_number);
+        System::<T>::initialize(&block_number, &parent_hash, &Default::default());
+        <crate::Pallet<T> as Hooks<T::BlockNumber>>::on_initialize(block_number);
+        System::<T>::finalize();
+    }
+
+    #[test]
+    fn test_block_tree() {
```

Comment:
can we make this test more readable? It tries to test all the things in one test and very dense to actually track what is tested and what is not tested

Reply:
Okay, I can split it into multiple tests but it may bring quite some duplicate setup code in each test.

Reply:
This is how I did withdraw so I can test very specific scenaios. https://github.com/subspace/subspace/pull/1646/files#diff-bc31bd26a15ac3e6de45a5242e5fc7883136b699fa411a472d0d3eb45c50f0f9R775

Your test cases' mileage may very but I got minimal duplication for withdraws

Reply:
`test_block_tree` is split into 6 tests, should be more readable now.